### PR TITLE
First nullability work on metadata

### PR DESCRIPTION
**Pull request by roji**
Annotates Annotatable (haha), EntityType and related types.

No particular difficulties encountered, this was quite straightforward. A couple notes:

* I've annotated mutating methods (e.g. EntityType.SetPrimaryKey) to return nullable, because they return the results of conventions which could return null. Is this correct?
* I know we don't like TODO comments, but in some places I've left special TODO-NULLABLE for stuff to check/improve (this was the pattern used in the runtime repo).

### CODE REVIEW

File: src/EFCore/Metadata/Internal/EntityType.cs [hunk 1]
```diff
@@ -652,9 +654,10 @@ public virtual Key SetPrimaryKey([CanBeNull] Property property, ConfigurationSou
                 }
             }
 
-            if (!propertiesNullOrEmpty)
+            if (properties?.Count > 0)
             {
-                foreach (var property in newKey.Properties)
+                // TODO-NULLABLE: Potential bug if AddKey returns null above (because of convention)?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/EntityType.cs:659]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/EntityType.cs [hunk 2]
```diff
@@ -1031,7 +1034,8 @@ public virtual void OnForeignKeyUpdating([NotNull] ForeignKey foreignKey)
 
             removed = foreignKey.PrincipalKey.ReferencingForeignKeys.Remove(foreignKey);
             Check.DebugAssert(removed, "removed is false");
-            removed = foreignKey.PrincipalEntityType.DeclaredReferencingForeignKeys.Remove(foreignKey);
+            // TODO-NULLABLE: Potential bug?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/EntityType.cs:1037]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/EntityType.cs [hunk 3]
```diff
@@ -1904,12 +1908,13 @@ public virtual SkipNavigation RemoveSkipNavigation([NotNull] SkipNavigation navi
             removed = navigation.ForeignKey?.ReferencingSkipNavigations.Remove(navigation) ?? true;
             Check.DebugAssert(removed, "removed is false");
 
-            removed = navigation.TargetEntityType.DeclaredReferencingSkipNavigations.Remove(navigation);
+            // TODO-NULLABLE: Potential bug?
```
Comment (roji) [inline src/EFCore/Metadata/Internal/EntityType.cs:1911]: /cc @AndriySvyryd

File: src/EFCore/Metadata/Internal/EntityType.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/EntityType.cs:659]: yup, add `&& newKey != null` to the condition here

File: src/EFCore/Metadata/Internal/EntityType.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/EntityType.cs:1037]: No, `DeclaredReferencingForeignKeys` will never be `null` here unless there's a bug in `AddForeignKey`.

This and the next methods don't have the best names.

File: src/EFCore/Metadata/Internal/EntityType.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/EntityType.cs:1911]: No

File: src/Shared/NonCapturingLazyInitializer.cs [hunk 4]
```diff
@@ -45,13 +51,14 @@ internal static class NonCapturingLazyInitializer
         }
 
         public static TValue EnsureInitialized<TValue>(
-            [CanBeNull] ref TValue target,
+            [CanBeNull, CA.AllowNull] ref TValue? target,
```
Comment (AndriySvyryd) [inline src/Shared/NonCapturingLazyInitializer.cs:54]: Shouldn't this and the next one be `NotNull`?

Comment (roji) [reply src/Shared/NonCapturingLazyInitializer.cs:54]: Yep, I messed it up, thanks.

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**